packages/swing-store/src/importer.js

```diff
@@ -0,0 +1,148 @@
+  // everything). The DB records we add are marked as incomplete (as
+  // if they had been pruned locally), and can be populated later when
+  // the artifact is retrieved.
+
+  // While unlikely, the getExportData() protocol *is* allowed to
+  // deliver multiple values for the same key (last one wins), or use
+  // 'null' to delete a previously-defined key. So our first pass both
+  // installs the kvStore shadow records, and de-dups/deletes the
+  // metadata records into this Map.
+
+  const allMetadata = new Map();
+
+  for await (const [key, value] of exporter.getExportData()) {
+    const [tag] = key.split('.', 1);
+    if (tag === 'kv') {
+      // 'kv' keys contain individual kvStore entries
+      const subKey = key.substring(tag.length + 1);
+      if (value == null) {
+        // Note '==' rather than '===': any nullish value implies deletion
+        kernelStorage.kvStore.delete(subKey);
+      } else {
+        kernelStorage.kvStore.set(subKey, value);
+      }
+    } else if (value == null) {
+      allMetadata.delete(key);
+    } else {
+      allMetadata.set(key, value);
+    }
+  }
+
+  // Now take each metadata record and install the stub/pruned entry
+  // into the DB.
+
+  for (const [key, value] of allMetadata.entries()) {
+    const [tag, ...pieces] = key.split('.');
+    if (tag === 'bundle') {
+      const bundleID = pieces[0];
+      internal.bundleStore.addBundleRecord(bundleID);
+    } else if (tag === 'snapshot') {
+      const [_vatID, endPos] = pieces;
+      if (endPos !== 'current') {
+        internal.snapStore.addSnapshotRecord(JSON.parse(value));
+      }
+      // metadata['snapshot.v1.current'] = 'snapshot.v1.5' , i.e. it
+      // points to the name of the current artifact. We could
+      // conceivably remember this and compare it against the .inUse
+      // property of that record, but it's not worth the effort (we
+      // might encounter the records in either order).
+    } else if (tag === 'transcript') {
+      const metadata = JSON.parse(value);
+      if (!!key.endsWith('.current') !== !!metadata.isCurrent) {
+        throw Fail`transcript key ${key} mismatches metadata ${metadata}`;
+      }
+      internal.transcriptStore.addTranscriptSpanRecord(metadata);
+    } else {
+      Fail`unknown export-data type ${q(tag)} on import`;
+    }
+  }
+
+  // All the metadata is now installed, and we're prepared for
+  // artifacts. We walk `getArtifactNames()` and offer each one to the
+  // submodule, which ignores historical ones (unless
+  // 'includeHistorical' is true), and validates+accepts the
+  // rest. This is an initial import, so we don't need to check if we
+  // already have the data, but the submodule function is free to do
+  // that check if they want.
+
+  for await (const name of exporter.getArtifactNames()) {
+    const chunkProvider = () => exporter.getArtifact(name);
+    const dataProvider = async () => buffer(Readable.from(chunkProvider()));
```

Comment:
I find these names a bit confusing.
```suggestion
    const makeChunkIterator = () => exporter.getArtifact(name);
    const dataProvider = async () => buffer(Readable.from(makeChunkIterator()));
```
and I'm pretty sure that the `Readable.from` is unnecessary.
```suggestion
    const makeChunkIterator = () => exporter.getArtifact(name);
    const dataProvider = async () => buffer(makeChunkIterator());
```

Reply:
Agreed that `chunkProvider` isn't super clear. But I'll miss the parallelism between XProvider and YProvider (for chunks/data). The `dataProvider` could become.. `makeData`? Doesn't seem quite right.

I'll change it as you said and maybe some better name for `dataProvider` will come to me.

Reply:
looks like the Readable.from isn't necessary here, but is still necessary in snapStore/bundleStore where I call `.pipe()` on it